contrib/babelfishpg_tsql/src/pl_funcs-2.c

```diff
@@ -351,58 +352,111 @@ coalesce_typmod_hook_impl(const CoalesceExpr *cexpr)
 }
 
+/*
+ * Check if the given object is in the restricted list
+ */
+static bool
+is_object_restricted(const char *objname, ObjectType type, const char *schema_name)
+{
+	const RestrictedObject	*restricted_objects;
+	size_t num_restricted_objects;
+
+	/* Determine which array to check based on schema */
+	if (pg_strcasecmp(schema_name, "master_dbo") == 0)
+	{
+		restricted_objects = master_dbo_objects;
+		num_restricted_objects = sizeof(master_dbo_objects)/sizeof(master_dbo_objects[0]);
+	}
+	else if (pg_strcasecmp(schema_name, "msdb_dbo") == 0)
+	{
+		restricted_objects = msdb_dbo_objects;
+		num_restricted_objects = sizeof(msdb_dbo_objects)/sizeof(msdb_dbo_objects[0]);
+	}
+	else
+		return false;
+
+	for (int idx = 0; idx < num_restricted_objects; idx++)
+	{
+		if (type == restricted_objects[idx].type &&
+			pg_strcasecmp(objname, restricted_objects[idx].name) == 0)
+			return true;
+	}
+
+	return false;
+}
+
+/*
+ * Main function to check and error out for restricted objects
+ */
 void
-check_restricted_stored_procedure(Oid proc_id)
+check_restricted_object(Oid object_id, ObjectType object_type)
 {
-	HeapTuple		proctup;
-	Form_pg_proc		procform;
-	const char		*procname;
-	Oid		schema_oid = InvalidOid;
-	Oid		dbo_oid = InvalidOid;
-	bool		is_restricted = false;
+	HeapTuple	tuple;
+	const char	*objname;
+	const char	*schema_name;
+	Oid			schema_oid;
 
-	/*
-	 * List of procedure names that are not allowed to be dropped. These procedures 
-	 * are considered essential or restricted due to security or operational reasons.
-	 */
-	static const char	*restricted_procedures[] = {
-		"xp_qv",
-		"xp_instance_regread",
-		"sp_addlinkedsrvlogin",
-		"sp_droplinkedsrvlogin",
-		"sp_dropserver",
-		"sp_enum_oledb_providers",
-		"sp_testlinkedserver"
-	};
-	static const int	RESTRICTED_PROCEDURES_COUNT = sizeof(restricted_procedures) / sizeof(restricted_procedures[0]);
-
-	proctup = SearchSysCache1(PROCOID, ObjectIdGetDatum(proc_id));
-
-	procform = (Form_pg_proc) GETSTRUCT(proctup);
-	procname = pstrdup(NameStr(procform->proname));
-	schema_oid = procform->pronamespace;
-	dbo_oid = get_namespace_oid("master_dbo", true);
-
-	if (OidIsValid(schema_oid) && OidIsValid(dbo_oid) && schema_oid == dbo_oid)
+	/* Get object information */
+	switch (object_type)
 	{
-		/* Check if the procedure name is in the restricted list */
-		for (int i = 0; i < RESTRICTED_PROCEDURES_COUNT; i++)
-		{
-			if (pg_strcasecmp(procname, restricted_procedures[i]) == 0)
-			{
-				is_restricted = true;
-				break;
-			}
-		}
+		case OBJECT_PROCEDURE:
+			tuple = SearchSysCache1(PROCOID, ObjectIdGetDatum(object_id));
+			break;
+		case OBJECT_VIEW:
+			tuple = SearchSysCache1(RELOID, ObjectIdGetDatum(object_id));
+			break;
+		case OBJECT_FUNCTION:
+			tuple = SearchSysCache1(PROCOID, ObjectIdGetDatum(object_id));
+			break;
+		default:
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("unsupported object type")));
+			return;
 	}
 
```

Comment:
You don't need to release the tuple here since you're going to raise an error and error handler will handle that.